lib/zookeeper.js

```diff
@@ -1,106 +1,331 @@
+        super();
+
+        if (_.isString(config)) {
+            this.config = { connect: config };
+        } else {
+            this.config = config;
+        }
+
+        this.native = new NativeZk();
+        this.native.emit = (ev, a1, a2, a3) => {
+            this.log(`Emitting '${ev}' with args: ${a1}, ${a2}, ${a3}`);
+            if (ev === 'connect' || ev === 'close') {
+                // the event is passing the native object. need to mangle to return the wrapper
+                a1 = this; // eslint-disable-line no-param-reassign
+            }
+            this.emit(ev, a1, a2, a3);
+        };
+
+        this.encoding = null;
+    }
+
+    setLogger(logger) {
+        if (logger === true) {
+            this.logger = (str) => {
+                console.info(`ZOOKEEPER_LOG: ${str}`);
+            };
+        } else if (logger === false) {
+            this.logger = undefined;
+        } else if (_.isFunction(logger)) {
+            this.logger = logger;
+        } else {
+            throw new Error('InvalidArgument: logger must be a function or true/false to utilize default logger');
+        }
+    }
+
+    log(message) {
+        if (this.logger) this.logger(message);
+    }
+
+    init(config) {
+        if (_.isString(config)) {
+            // eslint-disable-next-line no-param-reassign
+            config = { connect: config };
+        }
+        if (this.config) {
+            // eslint-disable-next-line no-param-reassign
+            config = config ? _.defaults(config, this.config) : this.config;
+        }
+        this.log(`Calling init with ${util.inspect(config)}`);
+        if (!_.isUndefined(config.data_as_buffer)) {
+            this.data_as_buffer = config.data_as_buffer;
+            this.log('Encoding for data output: %s', this.encoding);
+        }
+        this.native.init(config);
+
+        return this;
+    }
+
+    connect(options, cb) {
+        let callback;
+        if (_.isFunction(options)) {
+            callback = options;
+            this.init({});
+        } else {
+            callback = cb;
+            this.init(options);
+        }
+
+        this.errorHandler = (err) => {
+            this.removeListener('error', this.errorHandler);
+            this.removeListener('connect', this.connectHandler);
+            callback(err);
+        };
+
+        this.connectHandler = () => {
+            this.removeListener('error', this.errorHandler);
+            this.removeListener('connect', this.connectHandler);
+            callback(null, this);
+        };
+
+        this.on('error', this.errorHandler);
+        this.on('connect', this.connectHandler);
+    }
+
+    close() {
+        this.log(`Calling close with ${util.inspect([])}`);
+        return this.native.close();
+    }
+
+    a_create(path, data, flags, pathCb) {
+        this.log(`Calling a_create with ${util.inspect([path, data, flags, pathCb])}`);
+        return this.native.a_create(path, data, flags, pathCb);
```

Comment:
Very nice! I think I want to test this, to learn more about it, because I thought the replaced `apply` call passed in the args as an array to the native object. Or, should the arguments to `a_create` be an array?

Reply:
From what I recall, you can call the "native" methods normally. `apply` however expects arguments to be passed as array :wink:

Reply:
Got it! The different approaches with call and apply in the Master branch got me confused.